internal/devconfig/config.go

```diff
@@ -104,19 +114,36 @@ func loadBytes(b []byte) (*Config, error) {
 }
 
 func (c *Config) LoadRecursive(lockfile *lock.File) error {
+	return c.loadRecursive(lockfile, map[string]bool{})
+}
+
+func (c *Config) loadRecursive(
+	lockfile *lock.File,
+	seen map[string]bool,
+) error {
 	included := make([]*Config, 0, len(c.Root.Include))
 
 	for _, includeRef := range c.Root.Include {
-		pluginConfig, err := plugin.LoadConfigFromInclude(includeRef, lockfile)
+		pluginConfig, err := plugin.LoadConfigFromInclude(
+			includeRef, lockfile, c.absRootPath)
 		if err != nil {
 			return errors.WithStack(err)
 		}
 
+		if seen[pluginConfig.Source.Hash()] {
+			return errors.Errorf("circular include detected: %s", includeRef)
+		}
+		seen[pluginConfig.Source.Hash()] = true
+
 		includable := &Config{
 			Root:       pluginConfig.ConfigFile,
 			pluginData: &pluginConfig.PluginOnlyData,
 		}
-		if err := includable.LoadRecursive(lockfile); err != nil {
+		if hasDir, ok := pluginConfig.Source.(*plugin.LocalPlugin); ok {
+			includable.absRootPath = filepath.Dir(hasDir.Path())
+		}
+
+		if err := includable.loadRecursive(lockfile, maps.Clone(seen)); err != nil {
```

Comment:
why do we need to clone the `seen` map?

Would recommend adding the explanation as a comment, since non-obvious to me. 

EDIT: since we do `maps.Clone` in multiple places, maybe the comment belongs in the function docblock of `loadRecursive` as a warning or note

Reply:
maps are reference types so if we don't clone them the same map is shared in all calls of this function (which would return false positives because it would find instances where the same plugin is imported by different paths)